### PR TITLE
algo: applyGivenRotations Distributed (GPU)

### LINKED ISSUE #717: Porting application of Given Rotations to GPU
In #675 the algorithm has been implemented for MC, but for GPU it still requires some adaption (e.g. custom communication kernels have to be implemented)

### PR DESCRIPTION
**Pull request by albestro**
Close #717

Apart from enabling GPU tests, the main required change for the algorithm has been to port the custom communication kernels for P2P send/recv of a column to work also with GPU memory (via a CPU temporary buffer).

TODO:
- [x] ~#692 (see [comment](https://github.com/eth-cscs/DLA-Future/pull/718#issuecomment-1331823261))~
- [x] ~#719 (see [comment](https://github.com/eth-cscs/DLA-Future/pull/718#issuecomment-1326150675))~ after last changes in that PR, probably not really needed to wait for it anymore
- [x] ~Add dependency for temporary buffer allocation~
- [x] ~Refactor temporary buffer on CPU used for GPU memory communication~
- [x] Check if the workaround for `transform<GPU>` with a nullary callable is still required

### CODE REVIEW

File: include/dlaf/eigensolver/tridiag_solver/rot.h
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/rot.h:257]: @msimberg do you want to mention any specific PR/issues here for future reference?

Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/rot.h:257]: Just that https://github.com/eth-cscs/DLA-Future/pull/697 should fix this :)

**Comment by albestro** [comment]
About the last open point in TOOD list
> Add dependency for temporary buffer allocation

Currently, the algorithm schedule an application of a rotation and `sync_wait`s on its completion before going on scheduling the next one. This fact ensures that communications cannot be scheduled all together beforehand, and with them, also temporary buffers for the GPU case. If this assumption will drop, it is relevant to highlight that there is nothing that would stop to schedule, and more importantly allocate, all of them together.

I think at the moment it might be too early to think about it considering current algorithm constraints, so I would just not care. Do you agree? @rasolca @msimberg @teonnik 

In case you agree, I think it is relevant to point that out for future reference.

**Review by msimberg — CHANGES_REQUESTED** [review]

File: include/dlaf/eigensolver/tridiag_solver/rot.h
Comment (albestro) [reply include/dlaf/eigensolver/tridiag_solver/rot.h:69]: I think it is more or less the same concern I stated in https://github.com/eth-cscs/DLA-Future/pull/718#issuecomment-1322268336. Is it?

Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/rot.h:69]: Indeed, it looks like it! Sorry for not seeing the connection.

I think this is fine as it is now, but a refactor would probably be a good idea. One potential way of "fixing" this is to have nested `let_value`s, with the outer one taking the communicator, and the inner taking the memory view. In that case the memory view is only going to be allocated when the pipeline allows communication to happen. However, by the time we get to that refactoring we might have refactored the communication functions to work more generically, not just with tiles and the problem might be fixed by that instead...

**Comment by albestro** [comment]
With @teonnik we have found an API mismatch that has been easily solved (#721).

That change should™ not be in conflict with this PR.

File: include/dlaf/eigensolver/tridiag_solver/rot.h
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/rot.h:73]: I think this is ok:
```suggestion
                 ex::just(mem_view(), col_data, to_sizet(n) * sizeof(T),
```
None of the arguments look like senders to me. This very slightly reduces template instantiations (`just` is cheaper than `when_all`).

File: include/dlaf/eigensolver/tridiag_solver/rot.h
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/rot.h:257]: Conversely, I think all arguments are senders here:
```suggestion
          ex::when_all(ex::just(1),  // workaround for problem with unwrapping
```
Alternatively:
```suggestion
          di::whenAllLift(1,  // workaround for problem with unwrapping
```

**Review by msimberg — APPROVED** [review]
Two more minor style comments. Sorry for missing these earlier.

**Comment by teonnik** [comment]
LGTM!

Note that the PR may conflict with changes to rot.h in #719 : [include/dlaf/eigensolver/tridiag_solver/rot.h](https://github.com/eth-cscs/DLA-Future/pull/719/files#diff-745638b97d24e04ee5c3094a75fac7810a334a4b4c790afcc395a7d90639dcc7L90-R124)

**Comment by albestro** [comment]
> LGTM!
> 
> Note that the PR may conflict with changes to rot.h in #719 : [include/dlaf/eigensolver/tridiag_solver/rot.h](https://github.com/eth-cscs/DLA-Future/pull/719/files#diff-745638b97d24e04ee5c3094a75fac7810a334a4b4c790afcc395a7d90639dcc7L90-R124)

Yes, I will take care of that once your one gets merged.
Thanks for highlighting that 😉

**Review by msimberg — APPROVED** [review]

**Comment by albestro** [comment]
> ## try
> Build failed:
> 
> * [ci/gitlab/full-pipeline](https://gitlab.com/cscs-ci/eth-cscs/DLA-Future/-/pipelines/708238331)

Everything is ok, apart ROCm build because of the `givensRotationOnDevice` kernel.

I would go for changing all guards to `DLAF_WITH_CUDA` because of this

https://github.com/eth-cscs/DLA-Future/blob/1332ce2e576772cd537bb7359d9a7dd3eed5d0fd/src/CMakeLists.txt#L208-L217

waiting for your feedback before proceeding.

@rasolca @teonnik @msimberg @aurianer

**Comment by msimberg** [comment]
> Everything is ok, apart ROCm build because of the `givensRotationOnDevice` kernel.
> 
> I would go for changing all guards to `DLAF_WITH_CUDA` because of this
> 
> https://github.com/eth-cscs/DLA-Future/blob/1332ce2e576772cd537bb7359d9a7dd3eed5d0fd/src/CMakeLists.txt#L208-L217
> 
> waiting for your feedback before proceeding.

Slightly unsure since I don't know the extent of the missing HIP support in this PR, but would it make sense to make sure #692 is first merged, and then do whatever remaining adjustments here to make it work with HIP as well? Regular CUDA kernels should "just work" with HIP anyway so the question is whether this PR requires CUDA library calls that are not trivially translatable to HIP? Outside of the ones that have been updated in #692 of course.

**Comment by albestro** [comment]
bors try

@rasolca as per agreement I unlock this from HIP.

@teonnik as soon as this gets merged, you can proceed with another PR adapting tridiag solver to use this one.

There is the open question about https://github.com/eth-cscs/DLA-Future/pull/718#issuecomment-1322268336 for which I ask all of you @msimberg @teonnik @rasolca if you prefer to have an issue tracking that or it is ok leaving it there like it is.

**Comment by msimberg** [comment]
> bors try
> 
> @rasolca as per agreement I unlock this from HIP.
> 
> @teonnik as soon as this gets merged, you can proceed with another PR adapting tridiag solver to use this one.
> 
> There is the open question about [#718 (comment)](https://github.com/eth-cscs/DLA-Future/pull/718#issuecomment-1322268336) for which I ask all of you @msimberg @teonnik @rasolca if you prefer to have an issue tracking that or it is ok leaving it there like it is.

I guess an issue wouldn't hurt, unless you think it's trivial to remember to come back to it later...

**Comment by albestro** [comment]
@rasolca this is ready for the last bors run after:
- adding note about https://github.com/eth-cscs/DLA-Future/pull/718#issuecomment-1322268336
- https://github.com/eth-cscs/DLA-Future/pull/718/commits/62e98054d3d91427cdfd5b9d021f5b7818ad5c02 revert of the workaround thanks to the new unwrapping (it required a rebase onto new master)

I'm going to mark this as RTMWTP and I give to you full control for calling the last bors 😉

**Pull request merged**
